### PR TITLE
Split API dir into a separate sub module

### PR DESCRIPTION
**Pull request by filanov**
**Is this a BUG FIX or a FEATURE ?**:

/kind enhancement

**What this PR does / why we need it**:

Anyone who wants to vedor APIs into their project can suffer a dependecy hell.
The solution is to use submodels in the prokect to separate the API form the rest of the code and the imports.

**Special notes for your reviewer**:

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by filanov** [comment]
This come to resolve https://github.com/nmstate/kubernetes-nmstate/issues/864

**Comment by qinqon** [comment]
@filanov can you add Signed-Off-by with `git commit -s`

**Comment by phoracek** [comment]
@filanov would you please fix the vendoring directory. Per CI:

```
	To sync the vendor directory, run:
		go mod vendor
```

**Comment by qinqon** [comment]
@filanov do we need a vendor directory at api level too ?

File: api/go.mod [hunk 1]
```diff
@@ -0,0 +1,15 @@
+module github.com/nmstate/kubernetes-nmstate/api
+
+go 1.16
+
+require (
+	github.com/nmstate/kubernetes-nmstate/pkg/names v0.0.0
+	github.com/onsi/ginkgo v1.16.5
+	github.com/onsi/gomega v1.17.0
+	k8s.io/api v0.22.3
```
Comment (qinqon) [inline api/go.mod:9]: This is not aligend with main module https://github.com/nmstate/kubernetes-nmstate/blob/main/go.mod#L27

**Review by qinqon — CHANGES_REQUESTED** [review]
the api submodule has to pin the same k8s version as the main module.

**Comment by phoracek** [comment]
@qinqon I think we do. I tried to change the version of Ginko used by sub-module and it was not reflected in top-level vendor/.

File: api/go.mod (hunk 1 above)
Comment (filanov) [reply api/go.mod:9]: it was updated in both places automatically by `go mod tidy`, is there a reason to have `k8s.io/api v0.21.1` ?

**Comment by filanov** [comment]
> @qinqon I think we do. I tried to change the version of Ginko used by sub-module and it was not reflected in top-level vendor/.

changed the makefile target to run go mod tidy and go mod vendor on the submodule directories

File: api/go.mod (hunk 1 above)
Comment (qinqon) [reply api/go.mod:9]: Right we are testing towards k8s 1.21.

Comment (qinqon) [reply api/go.mod:9]: Ok, let's go with it, CI is fine and we are going to move to 1.22 soon.

Comment (qinqon) [reply api/go.mod:9]: Looks like we have to "replace" to pin to the version:

```
+ GOARCH=amd64
+ go build -o /tmp/knmstate/kubernetes-nmstate/build/_output/bin//handler.manager.linux-amd64 ./cmd/handler
# k8s.io/client-go/applyconfigurations/apiserverinternal/v1alpha1
vendor/k8s.io/client-go/applyconfigurations/apiserverinternal/v1alpha1/storageversion.go:62:34: not enough arguments in call to managedfields.ExtractInto
	have (*v1alpha1.StorageVersion, typed.ParseableType, string, *StorageVersionApplyConfiguration)
	want (runtime.Object, typed.ParseableType, string, interface {}, string)
# k8s.io/client-go/applyconfigurations/policy/v1

```

**Comment by qinqon** [comment]
@filanov something is not fine, CI is failing building the thing, looks like API version mismatch.

**Comment by filanov** [comment]
/hold 
will try to resolve it.

**Comment by filanov** [comment]
/unhold 

local build passed, looks like unit tests are passing as well

**Comment by qinqon** [comment]
Ther are still errors:
```
PASS
Failed to compile v1alpha1:

main module (github.com/nmstate/kubernetes-nmstate) does not contain package github.com/nmstate/kubernetes-nmstate/api/v1alpha1

There were failures detected in the following suites:
	 v1beta1 ./api/v1beta1
	   names ./pkg/names
	  shared ./api/shared
	v1alpha1 ./api/v1alpha1

Ginkgo ran 14 suites in 7m55.882685739s
Test Suite Failed
```

Can you run unit test locally first ?

**Comment by qinqon** [comment]
@filanov can you rebase?

**Comment by filanov** [comment]
> @filanov can you rebase?

Need to fix the tests, something related to gingko version, will try to get to it this week

**Comment by filanov** [comment]
In order to make the unit tests pass i need to run them separately for each submodel, but for some reason there is a problem to locate ginkgo package
```
# hande unit test for submodule
(cd ./pkg/names ; NODE_NAME=node01 go run github.com/onsi/ginkgo/ginkgo -r -keepGoing --randomizeAllSpecs --randomizeSuites --race --trace  ./...)
cannot find package "." in:
	/home/mfilanov/work/kubernetes-nmstate/pkg/names/vendor/github.com/onsi/ginkgo/ginkgo
make: *** [Makefile:170: test-unit-submodules-./pkg/names] Error 1
```
Any idea how it is installed?

**Comment by filanov** [comment]
Closing this PR, sadly don't have time to refactor it.

**Comment by filanov** [comment]
replaced by https://github.com/nmstate/kubernetes-nmstate/pull/911

**Pull request closed without merging**